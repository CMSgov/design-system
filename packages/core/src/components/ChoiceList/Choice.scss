@import '@cmsgov/design-system-support/src/settings/index';

/*
Checkbox & Radio

Checkboxes allow users to select one or more options from a visible list, whereas radio buttons allow a user to select only one option.

@uswds https://standards.usa.gov/components/form-controls

Markup: checkbox-radio.example.html

Style guide: components.choice
*/

/*
Inverse theme

Markup: checkbox-radio-inverse.example.html

Style guide: components.choice.inversed
*/

$ds-c-inset-border-width: $spacer-half;

// Hide the default browser checkbox/radio button since we'll
// create our own custom version
.ds-c-choice {
  left: -999em;
  opacity: 0;
  position: absolute;

  // Checked children container
  &__checkedChild {
    border-left: $ds-c-inset-border-width solid $color-primary;
    margin-bottom: $spacer-2;
    margin-left: ($choice-size / 2) - ($ds-c-inset-border-width / 2);
    padding: $spacer-2;

    // Checked children container on dark background
    &--inverse {
      border-left-color: $color-white;
    }

    // Small input variant
    &--small {
      margin-left: $spacer-1;
    }
  }
}

.ds-c-choice + label {
  align-items: center;
  cursor: pointer;
  display: flex;
  flex-wrap: wrap;
  font-weight: $font-normal;
  margin: $spacer-1 0;
  max-width: $measure-base;
  min-height: $choice-size;
  padding-left: $choice-size + $spacer-1;
  position: relative;
}

// Create a custom checkbox/radio button
.ds-c-choice + label::before {
  background-color: $color-background;
  border: $choice-border-width solid $choice-border-color;
  box-sizing: border-box;
  content: '\a0';
  height: $choice-size;
  left: 0;
  line-height: $choice-size;
  position: absolute;
  text-indent: 0.15em;
  top: 0;
  width: $choice-size;
}

.ds-c-choice--inverse + label::before {
  background-color: $color-background-inverse;
  border-color: $choice-border-color-inverse;
}

// Display an outline
.ds-c-choice:focus + label::before {
  box-shadow: 0 0 0 2px $color-background, 0 0 2px 4px $color-focus;
}

.ds-c-choice--inverse:focus + label::before {
  box-shadow: 0 0 0 2px $color-background-inverse,
    0 0 2px 4px $color-focus-inverse;
}

// Display a checkmark
.ds-c-choice:checked + label::before {
  background-color: $choice-checked-background-color;
  background-image: url('#{$image-path}/checkmark-white.svg');
  background-position: 50%;
  background-repeat: no-repeat;
  background-size: $choice-size - $spacer-1;
  border-color: $choice-checked-background-color;
}

// Fade the checkbox and label
.ds-c-choice:disabled {
  + label {
    color: $color-muted;

    &::before {
      background-color: $color-gray-lighter;
      border: 1px solid $color-gray-light;
      cursor: not-allowed;
    }
  }
}

.ds-c-choice--inverse:disabled {
  + label {
    color: $color-muted-inverse;

    &::before {
      background-color: rgba($color-muted-inverse, 0.15);
      box-shadow: 0 0 0 1px $color-muted-inverse;
    }
  }
}

// Display a circular radio button
.ds-c-choice[type='radio'] + label::before {
  border-radius: 100%;
}


/*
Right-to-Left

Markup: checkbox-radio-right-to-left.example.html

Style guide: components.choice.rtl
*/
.ds-c-choice--right + label {
  padding-left: 0;
  padding-right: $choice-size + $spacer-1;

  &::before {
    left: auto;
    right: 0;
  }
}

/*
Size variants

Markup: checkbox-radio-size.example.html

Style guide: components.choice.size
*/
.ds-c-choice--small {
  + label {
    min-height: $choice-size-small;
    padding-left: $choice-size-small + $spacer-1;
    padding-top: 0;

    &::before {
      height: $choice-size-small;
      top: 2px;
      width: $choice-size-small;
    }
  }

  &:checked + label::before {
    background-size: $choice-size-small;
  }

  &.ds-c-choice--right + label {
    padding-left: 0;
    padding-right: $choice-size-small + $spacer-1;
  }
}

/*
`<ChoiceList>`

@react-component ChoiceList

Style guide: components.choice.choicelist
*/

/*
`<Choice>`

@react-component Choice

Style guide: components.choice.react
*/

/*
---

## When to use

**Checkboxes**

- When a user can select any number of choices from a set list.
- When a user needs to choose “yes” or “no” on only one option (use a stand-alone checkbox). For example, to toggle a setting on or off.
- When users need to see all the available options at a glance.
- When users should be able to select zero of the options.

**Radio buttons**

- When users need to select only one option out of a set of mutually exclusive choices.

## When to consider alternatives

- If there are too many options to display on a mobile screen. Consider a `select` menu if you don’t have enough space to list out all available options, and if the user can only select one of the options.
- Never use radio buttons for optional questions, since once a radio button is selected from a list, it or another choice will remain selected.

## Guidance

- Don't rely on the visual difference between radio buttons and checkboxes. Make it clear with words when users can select one or multiple options.
- Users should be able to tap on or click on either the text label or the checkbox to select or deselect an option.
- In general, list choices vertically; horizontal listings can make it difficult to tell which label pertains to which choice. An exception is where you have binary choices with short labels, like 'Yes / No'. The convention here is for horizontal alignment.
- Avoid using negative language in labels as they can be counterintuitive. For example, “I want to receive a promotional email” instead of “I don’t want to receive promotional email.”
- Use caution if you decide to set a default value. Setting a default value can discourage users from making conscious decisions, seem pushy, or alienate users who don’t fit into your assumptions. In addition, you'll never know if the user explicitly chose that option or just didn't notice the question. If you're unsure, leave nothing selected by default.

**[View the "Forms" guidelines for additional guidance and best practices.](/guidelines/forms/)**

## Checked children and the expose within pattern

- The `<Choice>` component includes a `checkedChildren` prop that can expose hidden text information or form elements. This **expose within** pattern is especially useful if you need to collect data from follow up questions or give just-in-time feedback.
- Checked children can be exposed by checking the parent checkbox or radio button
- The `checkedChildren` prop should return one or more items wrapped in a `div` with the following className: `ds-c-choice__checkedChild`. This class sets the spacing and border color for the exposed elements.
- Add the className `ds-c-choice__checkedChild--inverse` to the `div` to show the inverse white border
- You may need to add the className `ds-u-margin--0` to your child element(s) to avoid extra top margin
- If you opt for smaller radio buttons or checkboxes, add className `ds-c-choice__checkedChild--small` to your checked child container

## Accessibility

- Surround a related set of choices with a `<fieldset>`. The `<legend>` provides context for the grouping. Do not use `fieldset` and `legend` for a single checkbox.
- Some screen readers read the `legend` text for each `fieldset`, so it should be brief and descriptive.
- Each input should have a semantic `id` attribute, and its corresponding `label` should have the same value in its `for` attribute.
- The custom checkboxes and radio buttons here are accessible to screen readers because the default fields are moved off-screen.
- `checkedChildren` will be announced to screen readers when they are exposed. They have been tested with the following devices:
  - Windows 10 + Internet Explorer 11 + JAWS screen reader
  - Windows 10 + Chrome + JAWS
  - Windows10 - Firefox + NVDA
    - NVDA reads out the `<select>` label and every `<option>` value
  - MacOS Mojave + Safari + VoiceOver

## Theming

The following Sass variables can be overridden to theme choice fields:

- `$choice-border-width`
- `$choice-border-color`
- `$choice-border-color-inverse`
- `$choice-checked-background-color`

## Related patterns

- [Select menu](/components/select)

## Learn more

- [Form Guidelines](/guidelines/forms/)
- [GOV.UK Checkbox/Radio buttons discussion](https://paper.dropbox.com/doc/Radio-buttons-oIwWoxwBKClt5IXvDbnpF)
- ["We've updated the radios and checkboxes on GOV.UK"](https://designnotes.blog.gov.uk/2016/11/30/weve-updated-the-radios-and-checkboxes-on-gov-uk/)
- [Four steps for choosing form elements on the Web (PDF)](http://www.formsthatwork.com/files/Articles/dropdown.pdf)

Style guide: components.choice.guidance
*/
