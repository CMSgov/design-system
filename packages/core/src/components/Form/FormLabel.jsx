import classNames from 'classnames';
import React from 'react';
import PropTypes from 'prop-types';

/**
 * The FormLabel component provides the label/legend for a field, along with any
 * associated hint text and error messaging.
 *
 * TODO(sawyer): Show this in the documentation
 */
export class FormLabel extends React.PureComponent {
  errorMessage() {
    if (this.props.errorMessage) {
      return (
        <span
          className='ds-c-field__hint ds-u-color--error'
          id={`${this.props.fieldId}-message`}
          role='alert'
        >
          {this.props.errorMessage}
        </span>
      );
    }
  }

  hint() {
    if (this.props.hint) {
      const classes = classNames(
        'ds-c-field__hint',
        {'ds-c-field__hint--inverse': this.props.inversed}
      );
      return <span className={classes}>{this.props.hint}</span>;
    }
  }

  render() {
    const ComponentType = this.props.component;
    const labelTextClasses = this.props.errorMessage && 'ds-u-font-weight--bold';

    return (
      <ComponentType className='ds-c-label' htmlFor={this.props.fieldId}>
        <span className={labelTextClasses}>{this.props.children}</span>
        {this.errorMessage()}
        {this.hint()}
      </ComponentType>
    );
  }
}

FormLabel.defaultProps = { component: 'label' };
FormLabel.propTypes = {
  children: PropTypes.node.isRequired,
  /** The root HTML element used to render the label */
  component: PropTypes.oneOf(['label', 'legend']),
  /** Enable the error state by providing an error message. */
  errorMessage: PropTypes.string,
  /**
   * The ID of the field this label is for. This is used for the label's `for`
   * attribute and any related ARIA attributes, such as for the error message.
   */
  fieldId: PropTypes.string,
  /**
   * Hint text. Typically this is a string, but you can pass in additional
   * HTML if you need to further format things.
   */
  hint: PropTypes.node,
  /**
   * Set to `true` to apply the "inverse" theme
   */
  inversed: PropTypes.bool
};

export default FormLabel;
