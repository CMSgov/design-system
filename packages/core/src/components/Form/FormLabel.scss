@import '../../settings/index';

/*
Form label/legend

Markup:
<label class="ds-c-label" for="field-id">Label text</label>
*/
.ds-c-label,
.ds-c-legend {
  display: block;
  font-size: $base-font-size;
  font-weight: $font-normal;
  margin-bottom: 0;
  margin-top: $spacer-3;
  max-width: $input-max-width;
}

.ds-c-legend {
  padding: 0;
}

/*
Hint text

(docs stolen from Gov.UK)

- donâ€™t use placeholder text in form fields, as this will disappear once content is entered into the form field
- use hint text for supporting contextual help, this will always be shown
- hint text should sit above a form field
- ensure hint text can be read correctly by screen readers
*/
.ds-c-field__hint {
  color: $color-muted;
  display: block;
}

.ds-c-field__hint--inverse {
  color: $color-muted-inverse;
}
