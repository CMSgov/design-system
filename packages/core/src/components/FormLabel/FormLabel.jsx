import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';

export class FormLabel extends React.PureComponent {
  constructor(props) {
    super(props);

    if (process.env.NODE_ENV !== 'production') {
      if (props.labelClassName) {
        console.warn(
          `[Deprecated]: Please remove the 'labelClassName' prop in <FormLabel>, use 'textClassName' instead. This prop has been renamed and will be removed in a future release.`
        );
      }
    }
  }
  errorMessage() {
    if (this.props.errorMessage) {
      const classes = classNames('ds-c-field__hint', 'ds-u-color--error', {
        'ds-u-color--error-light': this.props.inversed
      });

      return (
        <span className={classes} id={`${this.props.fieldId}-message`} role="alert">
          {this.props.errorMessage}
        </span>
      );
    }
  }

  hint() {
    const { hint } = this.props;
    let { requirementLabel } = this.props;
    if (!hint && !requirementLabel) return;

    const classes = classNames('ds-c-field__hint', {
      'ds-c-field__hint--inverse': this.props.inversed
    });

    let hintPadding = null;

    if (requirementLabel && hint) {
      if (typeof requirementLabel === 'string') {
        // Remove any existing spacing and punctuation
        requirementLabel = requirementLabel.trim().replace(/\.$/, '');
        // Add punctuation after the requirementLabel so it doesn't run into the hint
        requirementLabel = requirementLabel + '.';
      }

      // Add space between hint and preceding requirementLabel
      hintPadding = ' ';
    }

    return (
      <span className={classes}>
        {requirementLabel}
        {hintPadding}
        {hint}
      </span>
    );
  }

  render() {
    const { fieldId, id, children } = this.props;
    const ComponentType = this.props.component;
    const textClasses = classNames(this.props.labelClassName, this.props.textClassName);
    const classes = classNames('ds-c-label', this.props.className, {
      'ds-c-label--inverse': this.props.inversed
    });

    return (
      <ComponentType className={classes} htmlFor={fieldId} id={id}>
        <span className={textClasses}>{children}</span>
        {this.hint()}
        {this.errorMessage()}
      </ComponentType>
    );
  }
}

FormLabel.defaultProps = { component: 'label' };
FormLabel.propTypes = {
  /**
   * Label text or HTML.
   */
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  /**
   * Additional classes to be added to the root element.
   */
  className: PropTypes.string,
  /** The root HTML element used to render the label */
  component: PropTypes.oneOf(['label', 'legend']),
  /** Enable the error state by providing an error message. */
  errorMessage: PropTypes.node,
  /**
   * The ID of the field this label is for. This is used for the label's `for`
   * attribute and any related ARIA attributes, such as for the error message.
   */
  fieldId: PropTypes.string,
  /**
   * Additional hint text to display
   */
  hint: PropTypes.node,
  /**
   * A unique `id` for the label element. Useful for referencing the label from
   * other components with `aria-describedby`.
   */
  id: PropTypes.string,
  /**
   * Set to `true` to apply the "inverse" theme
   */
  inversed: PropTypes.bool,
  /**
   * @hide-prop [Deprecated] Additional classes to be added to the label text. Please use `textClassName` instead.
   */
  labelClassName: PropTypes.string,
  /**
   * Text showing the requirement (ie. "Optional", or "Required").
   * In most cases, this should be used to indicate which fields are optional.
   * See the [form guidelines]({{root}}/guidelines/forms/) for more info.
   */
  requirementLabel: PropTypes.node,
  /**
   * Additional classes to be added to the label text.
   */
  textClassName: PropTypes.string
};

export default FormLabel;
