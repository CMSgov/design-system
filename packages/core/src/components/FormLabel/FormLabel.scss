@import '@cmsgov/design-system-support/src/settings/index';

/*
Form Label & Legend

Form Labels & Legends are used with form inputs to provide labels, hints, and errors that help the user enter information.

Markup: form-label.example.html

Style guide: components.form-label
*/
.ds-c-label {
  display: block;
  font-size: $base-font-size;
  font-weight: $font-bold;
  margin-bottom: 0;
  margin-top: $spacer-3;
  max-width: $input-max-width;
  // Remove default padding on <legend> elements:
  padding: 0;
}

.ds-c-label--inverse {
  color: $color-base-inverse;
}

// Applies to "Checkbox & Radio" and "Date field"
.ds-c-fieldset {
  border: 0;
  // We apply the same margin as a label, since in the context of a fieldset
  // this would be a <legend>. See .ds-c-fieldset > .ds-c-label for more info.
  margin: $spacer-3 0 0;
  min-width: 0;
  padding: 0;
}

// Webkit won't render a top margin on a fieldset's <legend>, so we apply
// the top margin to the fieldset instead.
.ds-c-fieldset > .ds-c-label:first-child {
  margin-top: 0;
}

.ds-c-field__hint {
  color: $color-muted;
  display: block;
  font-weight: $font-normal;
}

.ds-c-field__hint--inverse {
  color: $color-muted-inverse;
}

.ds-c-choice + label .ds-c-field__hint {
  flex-basis: 100%;
}

/*
`<FormLabel>`

@react-component FormLabel

Style guide: components.form-label.react
*/

/*
---

## When to use

#### Labels

- Each form field should have a <label>. Never use a field's placeholder attribute as the primary way to label the field.

#### Legends

- Use a single legend for fieldset (this is required). One example of a common use of fieldset and legend is a question with radio button options for answers. The question text and radio buttons are wrapped in a fieldset, with the question itself being inside the legend tag.

## Guidance

#### Labels

- Apply the `ds-c-label` class to `<label>` elements.
- Each field should have a `<label>`. Never use a field's `placeholder` attribute as the primary way to label the field.
- Labels should have a `for` attribute, referencing the corresponding input's unique `id` attribute. Only one label can be associated to each unique form element.
- Labels should be placed above their fields.
- Label text should be short and in sentence case.
- Avoid colons at the end of labels.

#### Hint text

- Place hint text within the field's `<label>` element.
- Apply the `ds-c-field__hint` class to hint text.
- Use hint text for supporting contextual help, which will always be shown.
- Hint text should sit above a form field and below the label text.

#### Validation

- Place inline validation messages within the field's `<label>` element.
- Visually align inline validation messages with the input fields, so people using screen magnifiers can read them quickly.

## Learn More

- [Form Guidelines](/guidelines/forms/)

Style guide: components.form-label.guidance
*/
