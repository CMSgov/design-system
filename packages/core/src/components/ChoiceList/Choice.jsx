import EvEmitter from 'ev-emitter';
import FormLabel from '../FormLabel/FormLabel';
import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import uniqueId from 'lodash.uniqueid';

/** Used to emit events to all Choice components */
const dsChoiceEmitter = new EvEmitter();

export class Choice extends React.PureComponent {
  constructor(props) {
    super(props);

    if (process.env.NODE_ENV !== 'production') {
      if (props.inputPlacement === 'right') {
        console.warn(
          `[Deprecated]: Please remove the 'inputPlacement' prop in <Choice>. This prop is no longer supported and will be removed in a future release.`
        );
      }
    }

    this.input = null;
    this.handleChange = this.handleChange.bind(this);
    this.id = this.props.id || uniqueId(`${this.props.type}_${this.props.name}_`);

    if (typeof this.props.checked === 'undefined') {
      this.isControlled = false;
      // Since this isn't a controlled component, we need a way
      // to track when the value has changed. This can then be used
      // to identify when to toggle the visibility of (un)checkedChildren
      this.state = { checked: this.props.defaultChecked };

      // Event emitters are only relevant for uncontrolled radio buttons
      if (this.props.type === 'radio') {
        this.uncheckEventName = `${this.props.name}-uncheck`;
        dsChoiceEmitter.on(this.uncheckEventName, this.handleUncheck.bind(this));
      }
    } else {
      this.isControlled = true;
    }
  }

  checked() {
    if (this.isControlled) {
      return this.props.checked;
    }

    return this.state.checked;
  }

  /**
   * A radio button doesn't receive an onChange event when it is unchecked,
   * so we fire an "uncheck" event when any radio option is selected. This
   * allows us to check each radio options' checked state.
   * @param {String} checkedId - ID of the checked radio option
   */
  handleUncheck(checkedId) {
    if (checkedId !== this.id && this.input.checked !== this.state.checked) {
      this.setState({ checked: this.input.checked });
    }
  }

  handleChange(evt) {
    if (this.props.onChange) {
      this.props.onChange(evt);
    }

    if (!this.isControlled) {
      this.setState({ checked: evt.target.checked });
    }

    if (this.uncheckEventName && evt.target.checked) {
      // Emit the uncheck event so other radio options update their state
      dsChoiceEmitter.emitEvent(this.uncheckEventName, [this.id]);
    }
  }

  render() {
    const {
      checkedChildren,
      children,
      className,
      hint,
      inversed,
      inputPlacement,
      inputClassName,
      labelClassName,
      requirementLabel,
      size,
      uncheckedChildren,
      inputRef,
      ...inputProps
    } = this.props;

    const inputClasses = classNames(inputClassName, 'ds-c-choice', {
      'ds-c-choice--inverse': inversed,
      'ds-c-choice--right': inputPlacement === 'right',
      'ds-c-choice--small': size === 'small'
    });

    // Remove props we have our own implementations for
    if (inputProps.id) delete inputProps.id;
    if (inputProps.onChange) delete inputProps.onChange;

    return (
      <div
        className={className}
        aria-live={checkedChildren ? 'polite' : null}
        aria-relevant={checkedChildren ? 'additions text' : null}
        aria-atomic={checkedChildren ? 'false' : null}
      >
        <input
          className={inputClasses}
          id={this.id}
          onChange={this.handleChange}
          ref={ref => {
            this.input = ref;
            if (inputRef) {
              inputRef(ref);
            }
          }}
          {...inputProps}
        />
        <FormLabel
          className={labelClassName}
          fieldId={this.id}
          hint={hint}
          requirementLabel={requirementLabel}
        >
          {children}
        </FormLabel>
        {this.checked() ? checkedChildren : uncheckedChildren}
      </div>
    );
  }
}

Choice.defaultProps = {
  type: 'checkbox',
  inputPlacement: 'left'
};

Choice.propTypes = {
  /**
   * Label text or HTML.
   */
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  /**
   * Sets the input's `checked` state. Use this in combination with `onChange`
   * for a controlled component; otherwise, set `defaultChecked`.
   */
  checked: PropTypes.bool,
  /**
   * Content to be shown when the choice is checked. See
   * **Checked children and the expose within pattern** on
   * the Guidance tab for detailed instructions.
   */
  checkedChildren: PropTypes.node,
  /**
   * Content to be shown when the choice is not checked
   */
  uncheckedChildren: PropTypes.node,
  /**
   * Additional classes to be added to the root `div` element.
   */
  className: PropTypes.string,
  /**
   * Additional classes to be added to the `input` element.
   */
  inputClassName: PropTypes.string,
  /**
   * Additional classes to be added to the `FormLabel`.
   */
  labelClassName: PropTypes.string,
  /**
   * Sets the initial `checked` state. Use this for an uncontrolled component;
   * otherwise, use the `checked` property.
   */
  defaultChecked: PropTypes.bool,
  /**
   * Access a reference to the `input` element
   */
  inputRef: PropTypes.func,
  /**
   * Additional hint text to display below the choice's label
   */
  hint: PropTypes.node,
  /**
   * A unique ID to be used for the input field, as well as the label's
   * `for` attribute. A unique ID will be generated if one isn't provided.
   */
  id: PropTypes.string,
  /**
   * Text showing the requirement ("Required", "Optional", etc.). See [Required and Optional Fields]({{root}}/guidelines/forms/#required-and-optional-fields).
   */
  requirementLabel: PropTypes.node,
  /**
   * Applies the "inverse" UI theme
   */
  inversed: PropTypes.bool,
  /**
   * @hide-prop [Deprecated] Placement of the input relative to the text label
   */
  inputPlacement: PropTypes.oneOf(['left', 'right']),
  size: PropTypes.oneOf(['small']),
  /**
   * The `input` field's `name` attribute
   */
  name: PropTypes.string.isRequired,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  type: PropTypes.oneOf(['checkbox', 'radio']),
  /**
   * The `input` `value` attribute
   */
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired
};

export default Choice;
