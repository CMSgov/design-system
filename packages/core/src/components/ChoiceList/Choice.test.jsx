import Choice from './Choice';
import FormLabel from '../FormLabel/FormLabel';
import React from 'react';
import { shallow } from 'enzyme';

const label = 'George Washington';

function shallowRender(customProps = {}, children = label) {
  const props = Object.assign(
    {
      name: 'foo',
      value: 'boo'
    },
    customProps
  );

  return {
    props: props,
    wrapper: shallow(<Choice {...props}>{children}</Choice>)
  };
}

describe('Choice', () => {
  it('accepts a node as innerHTML', () => {
    const wrapper = shallowRender(
      {},
      <p>
        <strong>Hello</strong> World
      </p>
    ).wrapper;
    const labelNode = wrapper.find(FormLabel).dive();

    expect(
      labelNode
        .children()
        .first()
        .children()
        .first()
        .is('p')
    ).toBe(true);
  });

  it('is not checked', () => {
    const wrapper = shallowRender().wrapper;
    const input = wrapper.find('input');

    expect(input.prop('checked')).toBeUndefined();
    expect(input.prop('defaultChecked')).toBeUndefined();
  });

  it('is checked', () => {
    const wrapper = shallowRender({ checked: true }).wrapper;
    const input = wrapper.find('input');

    expect(input.prop('checked')).toBe(true);
    expect(input.prop('defaultChecked')).toBeUndefined();
  });

  it('is defaultChecked', () => {
    const wrapper = shallowRender({ defaultChecked: true }).wrapper;
    const input = wrapper.find('input');

    expect(input.prop('checked')).toBeUndefined();
    expect(input.prop('defaultChecked')).toBe(true);
  });

  it('is required', () => {
    const wrapper = shallowRender({ required: true }).wrapper;
    const input = wrapper.find('input');

    expect(input.prop('required')).toBe(true);
  });

  it('is a checkbox by default', () => {
    const wrapper = shallowRender().wrapper;
    const input = wrapper.find('input');

    expect(input.prop('type')).toBe('checkbox');
    expect(input.prop('required')).toBeUndefined();
  });

  it('is a radio button', () => {
    const wrapper = shallowRender({ type: 'radio' }).wrapper;
    const input = wrapper.find('input');

    expect(input.prop('type')).toBe('radio');
  });

  it('applies className to input', () => {
    const wrapper = shallowRender().wrapper;
    const input = wrapper.find('input');

    expect(input.hasClass('ds-c-choice')).toBe(true);
    expect(input.hasClass('ds-c-choice--inverse')).toBe(false);
  });

  it('has a hint and requirementLabel', () => {
    const wrapper = shallowRender({
      hint: 'Hello world',
      requirementLabel: 'Optional '
    }).wrapper;

    expect(wrapper.find('FormLabel')).toMatchSnapshot();
  });

  it('applies inverse className to input', () => {
    const wrapper = shallowRender({ inversed: true }).wrapper;
    const input = wrapper.find('input');

    expect(input.hasClass('ds-c-choice')).toBe(true);
    expect(input.hasClass('ds-c-choice--inverse')).toBe(true);
  });

  it('places input on left by default', () => {
    const wrapper = shallowRender().wrapper;
    const input = wrapper.find('input');

    expect(input.hasClass('ds-c-choice')).toBe(true);
    expect(input.hasClass('ds-c-choice--right')).toBe(false);
  });

  it('places input on right', () => {
    const wrapper = shallowRender({ inputPlacement: 'right' }).wrapper;
    const input = wrapper.find('input');

    expect(input.hasClass('ds-c-choice')).toBe(true);
    expect(input.hasClass('ds-c-choice--right')).toBe(true);
  });

  it('applies small className to input', () => {
    const wrapper = shallowRender({ size: 'small' }).wrapper;
    const input = wrapper.find('input');

    expect(input.hasClass('ds-c-choice')).toBe(true);
    expect(input.hasClass('ds-c-choice--small')).toBe(true);
  });

  it('applies additional classNames to root element', () => {
    const wrapper = shallowRender({ className: 'foo' }).wrapper;

    expect(wrapper.hasClass('foo')).toBe(true);
  });

  it('applies additional classNames to input element', () => {
    const wrapper = shallowRender({ inputClassName: 'foo' }).wrapper;
    const input = wrapper.find('input');

    expect(input.hasClass('foo')).toBe(true);
  });

  it('accepts a string value', () => {
    const data = shallowRender({ value: 'bar' });
    const input = data.wrapper.find('input');

    expect(input.prop('value')).toBe(data.props.value);
  });

  it('accepts a number value', () => {
    const data = shallowRender({ value: 100 });
    const input = data.wrapper.find('input');

    expect(input.prop('value')).toBe(data.props.value);
  });

  it('accepts a custom id', () => {
    const data = shallowRender({ id: 'custom_id' });
    const input = data.wrapper.find('input');
    const labelNode = data.wrapper.find(FormLabel).dive();

    expect(input.prop('id')).toBe(data.props.id);
    expect(labelNode.prop('htmlFor')).toBe(data.props.id);
  });

  it('generates a unique id', () => {
    const sharedProps = {
      name: 'presidents'
    };
    const wrapper = shallow(
      <div>
        <Choice value="a" {...sharedProps}>
          {label}
        </Choice>
        <Choice value="b" {...sharedProps}>
          {label}
        </Choice>
      </div>
    );
    const $wrapper = wrapper.render();

    const idRegex = new RegExp(`checkbox_${sharedProps.name}_[0-9]+`);
    const inputAId = $wrapper
      .find('input')
      .eq(0)
      .attr('id');
    const inputBId = $wrapper
      .find('input')
      .eq(1)
      .attr('id');

    // IDs should be unique!
    expect(inputAId).not.toBe(inputBId);

    // First Choice
    expect(inputAId).toMatch(idRegex);
    expect(
      $wrapper
        .find('label')
        .eq(0)
        .attr('for')
    ).toBe(inputAId);

    // Second choice
    expect(inputBId).toMatch(idRegex);
    expect(
      $wrapper
        .find('label')
        .eq(1)
        .attr('for')
    ).toBe(inputBId);
  });

  describe('state', () => {
    it('sets state for uncontrolled component', () => {
      const data = shallowRender({ defaultChecked: true });

      expect(data.wrapper.instance().isControlled).toBe(false);
      expect(data.wrapper.state('checked')).toBe(data.props.defaultChecked);
    });

    it('does not set state for controlled component', () => {
      const data = shallowRender({ checked: true });

      expect(data.wrapper.instance().isControlled).toBe(true);
      expect(data.wrapper.state('checked')).toBeUndefined();
    });
  });

  describe('event handlers and emitters', () => {
    let props;

    beforeEach(() => {
      props = {
        onBlur: jest.fn(),
        onChange: jest.fn()
      };
    });

    describe('onChange', () => {
      it('calls the onChange handler', () => {
        const data = shallowRender(props);
        const input = data.wrapper.find('input');

        input.simulate('change', {
          target: { checked: true }
        });

        expect(data.props.onBlur.mock.calls.length).toBe(0);
        expect(data.props.onChange.mock.calls.length).toBe(1);
      });

      it('updates state when uncontrolled component', () => {
        props.defaultChecked = true;
        const data = shallowRender(props);
        const input = data.wrapper.find('input');

        input.simulate('change', {
          target: { checked: false }
        });

        expect(data.wrapper.state('checked')).toBe(false);
      });

      it('skips updating state when controlled component', () => {
        props.checked = true;
        const data = shallowRender(props);
        const input = data.wrapper.find('input');

        input.simulate('change', {
          target: { checked: false }
        });

        expect(data.wrapper.state('checked')).toBeUndefined();
      });
    });

    it('calls the onBlur handler', () => {
      const data = shallowRender(props);
      const input = data.wrapper.find('input');

      input.simulate('blur');

      expect(data.props.onBlur.mock.calls.length).toBe(1);
      expect(data.props.onChange.mock.calls.length).toBe(0);
    });

    describe('uncheck event emitter', () => {
      it('sets uncheckEventName for uncontrolled radio buttons', () => {
        const data = shallowRender({ type: 'radio', defaultChecked: false });

        expect(data.wrapper.instance().uncheckEventName).toBe(
          `${data.props.name}-uncheck`
        );
      });

      it('does not set uncheckEventName for controlled radio buttons', () => {
        const data = shallowRender({ type: 'radio', checked: false });

        expect(data.wrapper.instance().uncheckEventName).toBeUndefined();
      });

      it('does not set uncheckEventName for uncontrolled checkbox', () => {
        const data = shallowRender({ defaultChecked: true });

        expect(data.wrapper.instance().uncheckEventName).toBeUndefined();
      });
    });
  });

  describe('nested content', () => {
    let props;

    function expectCheckedChildren(wrapper) {
      expect(wrapper.find('.unchecked-child').length).toBe(0);
      expect(wrapper.find('.checked-child').length).toBe(1);
    }

    function expectUncheckedChildren(wrapper) {
      expect(wrapper.find('.unchecked-child').length).toBe(1);
      expect(wrapper.find('.checked-child').length).toBe(0);
    }

    beforeEach(() => {
      props = {
        checkedChildren: (
          <strong className="checked-child">I am checked</strong>
        ),
        uncheckedChildren: (
          <strong className="unchecked-child">I am unchecked</strong>
        ),
        name: 'foo',
        value: 'bar'
      };
    });

    describe('controlled component', () => {
      it('renders uncheckedChildren when not checked', () => {
        const wrapper = shallow(<Choice {...props}>Foo</Choice>);

        expectUncheckedChildren(wrapper);
      });

      it('renders uncheckedChildren when checked is changed to false', () => {
        props.checked = true;
        const wrapper = shallow(<Choice {...props}>Foo</Choice>);
        wrapper.setProps({ checked: false });

        expectUncheckedChildren(wrapper);
      });

      it('renders checkedChildren when checked', () => {
        props.checked = true;
        const wrapper = shallow(<Choice {...props}>Foo</Choice>);

        expectCheckedChildren(wrapper);
      });

      it('renders checkedChildren when checked is changed to true', () => {
        props.checked = false;
        const wrapper = shallow(<Choice {...props}>Foo</Choice>);
        wrapper.setProps({ checked: true });

        expectCheckedChildren(wrapper);
      });
    });

    describe('uncontrolled component', () => {
      it('renders uncheckedChildren when not defaultChecked', () => {
        props.defaultChecked = false;
        const wrapper = shallow(<Choice {...props}>Foo</Choice>);

        expectUncheckedChildren(wrapper);
      });

      it('renders uncheckedChildren when changed to unchecked', () => {
        props.defaultChecked = true;
        const wrapper = shallow(<Choice {...props}>Foo</Choice>);

        wrapper.setState({ checked: false });
        expectUncheckedChildren(wrapper);
      });

      it('renders checkedChildren when defaultChecked', () => {
        props.defaultChecked = true;
        const wrapper = shallow(<Choice {...props}>Foo</Choice>);

        expectCheckedChildren(wrapper);
      });
    });
  });
});
