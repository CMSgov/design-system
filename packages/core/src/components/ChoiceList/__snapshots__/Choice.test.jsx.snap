// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Choice has a hint and requirementLabel 1`] = `
<FormLabel
  component="label"
  fieldId="checkbox_foo_9"
  hint="Hello world"
  requirementLabel="Optional "
>
  George Washington
</FormLabel>
`;
